sccache-dist/src/cache.rs

```diff
@@ -0,0 +1,158 @@
+        }
+        let weak_map = serde_json::from_reader(fs::File::open(weak_map_path).unwrap()).unwrap();
+
+        let tc_cache_dir = cache_dir.join("tc");
+        let cache = Mutex::new(TcCache::new(&tc_cache_dir, cache_size).unwrap());
+
+        Self {
+            cache_dir,
+            cache,
+            // TODO: shouldn't clear on restart, but also should have some
+            // form of pruning
+            weak_map: Mutex::new(weak_map),
+        }
+    }
+
+    // Get the bytes of a toolchain tar
+    pub fn get_toolchain_cache(&self, key: &str) -> Option<Vec<u8>> {
+        let mut toolchain_reader = match self.cache.lock().unwrap().get(key) {
+            Ok(rdr) => rdr,
+            Err(LruError::FileNotInCache) => return None,
+            Err(e) => panic!("{}", e),
+        };
+        let mut ret = vec![];
+        toolchain_reader.read_to_end(&mut ret).unwrap();
+        Some(ret)
+    }
+    // TODO: It's more correct to have a FnBox or Box<FnOnce> here
+    // If the toolchain doesn't already exist, create it and insert into the cache
+    pub fn put_toolchain_cache(&self, weak_key: &str, create: &mut FnMut(fs::File) -> io::Result<()>) -> Result<String> {
+        if let Some(strong_key) = self.weak_to_strong(weak_key) {
+            debug!("Using cached toolchain {} -> {}", weak_key, strong_key);
+            return Ok(strong_key)
+        }
+        debug!("Weak key {} appears to be new", weak_key);
+        // TODO: don't use this as a poor exclusive lock on this global file
+        let mut cache = self.cache.lock().unwrap();
+        let file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open("/tmp/toolchain_cache.tar")?;
+        create(file)?;
+        // TODO: after, if still exists, remove it
+        let strong_key = cache.insert_file("/tmp/toolchain_cache.tar")?;
+        self.record_weak(weak_key.to_owned(), strong_key.clone());
+        Ok(strong_key)
+    }
+
+    fn weak_to_strong(&self, weak_key: &str) -> Option<String> {
+        self.weak_map.lock().unwrap().get(weak_key).map(String::to_owned)
+    }
+    fn record_weak(&self, weak_key: String, key: String) {
+        let mut weak_map = self.weak_map.lock().unwrap();
+        weak_map.insert(weak_key, key);
+        let weak_map_path = self.cache_dir.join("weak_map.json");
+        serde_json::to_writer(fs::File::create(weak_map_path).unwrap(), &*weak_map).unwrap()
+    }
+}
+
+// Partially copied from util.rs
+fn hash_reader<R: Read + Send + 'static>(rdr: R) -> Result<String> {
+    let mut m = Context::new(&SHA512);
+    let mut reader = BufReader::new(rdr);
+    loop {
+        let mut buffer = [0; 1024];
+        let count = reader.read(&mut buffer[..])?;
+        if count == 0 {
+            break;
+        }
+        m.update(&buffer[..count]);
+    }
+    Ok(hex(m.finish().as_ref()))
+}
+
+fn hex(bytes: &[u8]) -> String {
```

Comment:
This feels like something we could pull in from a crate, like the `hex` crate or similar.

Reply:
Oh, hah! If you move this code into the existing crate then you can just reuse that code, I guess. I'm certainly not going to hold you to a higher standard than the existing code. :)